keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,266 @@
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes a way to obtain service account token for pods that the CSI
+drivers are mounting volumes for. Since these tokens are valid only for a
+limited period, this KEP will also give the CSI drivers an option to re-execute
+`NodePublishVolume` to mount volumes.
+
+## Motivation
+
+Currently, the only way that CSI drivers acquire service account tokens is to
+directly read the token in the file system. However, this approach has
+uncharming traits:
+
+1.  It will not work for csi drivers which run as a different non-root user than
+    the pods. See
+    [file permission section for service account token](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20180515-svcacct-token-volumes.md#file-permission).
+2.  CSI driver will have access to the secrets of pods that do not use it
+    because the CSI driver should have a `hostPath` volume for the `pods`
+    subdirectory to read the token.
+3.  The audience of the token is defaulted to kube apiserver.
+4.  The token is not guaranteed to be available (e.g.
+    `automountServiceAccountToken=false`).
+
+### User stories
+
+- HashiCorp Vault provider for secret store CSI driver requires service
+  account token of the pods they are mounting secrets at to authenticate to
+  Vaults. The provisioned secrets also have given TTL in Vault, so it is
+  necessary get tokens after the initial mount.
+- Cert manager CSI dirver will create CertificateRequests on behalf of the
+  pods.
+- Amazon EFS CSI driver wants the service account tokens of pods to exchange
+  for AWS credentials.
+
+## Proposal
+
+### Goals
+
+- Allow CSI driver to request audience-bounded service account tokens of pods
+  from kubelet to `NodePublishVolume`.
+- Provide an option to re-execute `NodePublishVolume` in a best-effort manner.
+
+### Non-Goals
+
+- Other CSI calls may not acquire pods' service account tokens via this feature.
+- Failed re-execution of `NodePublishVolume` will not unmount volumes.
+
+### API Changes
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountToken *ServiceAccountToken
+}
+
+// ServiceAccountToken contains parameters of a token.
+type ServiceAccountToken struct {
+    Audiences []string
+    ExpirationSeconds *int64
+}
+```
+
+These three fields are all optional:
+
+- **`ServiceAccountToken.Audiences`**: will be set in `TokenRequestSpec`. All
+  the audiences will be validated to not be empty (i.e. kube apiserver). SP of
+  the CSI driver is supposed to send a `TokenReview` with at least one of the
+  audiences specified.
+
+- **`ServiceAccountToken.ExpirationSeconds`**: will be set in
+  `TokenRequestSpec`. The issued token may have a different duration, so the
+  `ExpirationTimestamp` in `TokenRequestStatus` will be passed to CSI driver.
+
+- **`RequiresRemount`**: should be only set when the mounted volumes by the
```

Comment:
We can discuss during the actual api review, but I don't think we should use the "remount" term since it's not actually a remount. Maybe republish or refresh? I think an important behavior here is don't call NodeUnpublish every time.

Reply:
yea, let's discuss the option in implementation because right now the name is consistent with what we have in kubelet. we should change it in all places.